velox/exec/fuzzer/WindowFuzzer.h

```diff
@@ -0,0 +1,184 @@
+      : AggregationFuzzerBase{seed, customVerificationFunctions, customInputGenerators, timestampPrecision, queryConfigs, std::move(referenceQueryRunner)},
+        orderDependentFunctions_{orderDependentFunctions} {
+    VELOX_CHECK(
+        !aggregationSignatureMap.empty() || !windowSignatureMap.empty(),
+        "No function signatures available.");
+
+    if (persistAndRunOnce_ && reproPersistPath_.empty()) {
+      std::cout
+          << "--repro_persist_path must be specified if --persist_and_run_once is specified"
+          << std::endl;
+      exit(1);
+    }
+
+    // Presto doesn't allow complex-typed vectors with NULL elements to be
+    // partition keys, so we disable it for all inputs.
+    // TODO: allow complex-typed vectors with NULL elements as non-key columns.
+    if (dynamic_cast<PrestoQueryRunner*>(referenceQueryRunner_.get())) {
+      auto options = vectorFuzzer_.getOptions();
+      options.containerHasNulls = false;
+      vectorFuzzer_.setOptions(options);
+    }
+
+    addAggregationSignatures(aggregationSignatureMap);
+    addWindowFunctionSignatures(windowSignatureMap);
+    printStats(functionsStats);
+
+    sortCallableSignatures(signatures_);
+    sortSignatureTemplates(signatureTemplates_);
+
+    signatureStats_.resize(signatures_.size() + signatureTemplates_.size());
+  }
+
+  void go();
+  void go(const std::string& planPath);
+
+ private:
+  struct SortingKeyAndOrder {
+    std::string key_;
+    std::string order_;
+    std::string nullsOrder_;
+
+    SortingKeyAndOrder() = delete;
+
+    SortingKeyAndOrder(
+        const std::string& key,
+        const std::string& order,
+        const std::string& nullsOrder) {
+      key_ = key;
+      order_ = order;
+      nullsOrder_ = nullsOrder;
+    }
+  };
+
+  void addWindowFunctionSignatures(const WindowFunctionMap& signatureMap);
+
+  const std::string generateFrameClause();
+
+  const std::string generateOrderByClause(
+      const std::vector<SortingKeyAndOrder>& sortingKeysAndOrders);
+
+  std::string getFrame(
+      const std::vector<std::string>& partitionKeys,
+      const std::vector<SortingKeyAndOrder>& sortingKeysAndOrders,
+      const std::string& frameClause);
+
+  std::vector<SortingKeyAndOrder> generateSortingKeysAndOrders(
+      const std::string& prefix,
+      std::vector<std::string>& names,
+      std::vector<TypePtr>& types);
+
+  void updateReferenceQueryStats(
+      AggregationFuzzerBase::ReferenceQueryErrorCode ec);
+
+  // Return 'true' if query plans failed.
+  bool verifyWindow(
+      const std::vector<std::string>& partitionKeys,
+      const std::vector<SortingKeyAndOrder>& sortingKeysAndOrders,
+      const std::string& frameClause,
+      const std::string& functionCall,
+      const std::vector<RowVectorPtr>& input,
+      bool customVerification,
+      bool enableWindowVerification);
+
+  const std::unordered_set<std::string> orderDependentFunctions_;
+
+  struct Stats {
```

Comment:
nit: since this is common to AggFuzzer too should we move this out to base ?

Reply:
AggregationFuzzer actually has a different set of Stats where some fields do not apply to window fuzzer. So I kept them separate when I extracted AggregationFuzzerBase. I think we can make a base type for Stats in AggregationFuzzerBase and move the common fields there. What do you think?

Reply:
Sounds good - we can just have one stats struct with all fields if it will make things easier ..

Reply:
I think having all fields in one stats struct where some of them are not used by individual fuzzers may confuse the fuzzer developers. Having the common fields in a base stats struct and let individual fuzzer extend it is better in my opinion.